### ISSUE TITLE
dtypes passes as strings don't work in array creation methods

### ISSUE THREAD

**Issue by seibert** [numba/numba#2201]
This kind of function does not work:
```python
@numba.jit(nopython=True)
def test_func():
    x = numpy.empty(100, dtype='int64')
    x[0] = 10
    return x
```
although `int64` is interpreted as a valid dtype by NumPy.

Could this be solved by creating a casting rule that allows strings to be cast to dtype objects?

**Comment by sklam** [comment]
> Could this be solved by creating a casting rule that allows strings to be cast to dtype objects?

better make an overload to take `dtype=str'` instead of new casting rule.

**Comment by seibert** [comment]
yeah, probably, though any function that takes a dtype object should probably do the same conversion

**Comment by stuartarchibald** [comment]
Also, with reference to #2205 

```python
from numba import cuda
@cuda.jit('void(float64[:])')
def kern(arr):
    cuda.local.array(10, 'float64')

kern[1,1](np.zeros(1))
```
gives:
```
numba.errors.LoweringError: Failed at nopython (nopython mode backend)
cannot create weak reference to 'str' object
File "issue2205.py", line 4
[1] During: lowering "$0.6 = call ptx.lmem.alloc($const0.4, $const0.5)" at issue2205.py (4)
```

**Comment by esc** [comment]
Since #6262 has now been merged, can this be closed? It seems like #6262 only partially addressed this issue, perhaps we could update here what is still missing, if anything.

**Comment by stuartarchibald** [comment]
> Since #6262 has now been merged, can this be closed? It seems like #6262 only partially addressed this issue, perhaps we could update here what is still missing, if anything.

Technically I think yes https://github.com/numba/numba/pull/6262#pullrequestreview-493577596

**Comment by stuartarchibald** [comment]
Updated https://github.com/numba/numba/issues/2205 as feature request. Suggest opening new tickets for any remaining issues.
